Source/WebKit2/UIProcess/API/wpe/WPEView.cpp

```diff
@@ -121,37 +150,136 @@ View::View(struct wpe_view_backend* backend, const API::PageConfiguration& baseC
     m_pageProxy->initializeWebPage();
 }
 
-View::~View()
+void View::initializeClient(const WKViewClientBase* client)
 {
-    wpe_view_backend_destroy(m_backend);
+    if (RunLoop::isMain())
+    {
+        m_client.initialize(client);
+    }
+    else
+    {
+        RunLoop::main().dispatch([this, client]
+        {
+            initializeClient(client);
+        });
+    }
 }
 
-void View::initializeClient(const WKViewClientBase* client)
+void View::frameDisplayed()
 {
-    m_client.initialize(client);
+    if (RunLoop::isMain())
+    {
+        m_client.frameDisplayed(*this);
+    }
+    else
+    {
+        RunLoop::main().dispatch([this]
+        {
+            frameDisplayed();
+        });
+    }
 }
 
-void View::frameDisplayed()
+void View::setSize(const WebCore::IntSize& size, bool recursive)
+{
+    if (!recursive)
+        m_size = size;
+
+    if (RunLoop::isMain())
+    {
+        if (page().drawingArea())
+        {
+            page().drawingArea()->setSize(size, WebCore::IntSize(), WebCore::IntSize());
+        }
+    }
+    else
+    {
+        RunLoop::main().dispatch([=]
+        {
+            setSize(size, true);
+        });
+    }
+}
+
+void View::handleKeyboardEvent(const WebKit::NativeWebKeyboardEvent& event)
 {
-    m_client.frameDisplayed(*this);
+    if (RunLoop::isMain())
+    {
+        page().handleKeyboardEvent(event);
+    }
+    else
+    {
+        RunLoop::main().dispatch([=]
+        {
+            handleKeyboardEvent(event);
+        });
+    }
 }
 
-void View::setSize(const WebCore::IntSize& size)
+void View::handleMouseEvent(const WebKit::NativeWebMouseEvent& event)
 {
-    m_size = size;
-    if (m_pageProxy->drawingArea())
-        m_pageProxy->drawingArea()->setSize(size, WebCore::IntSize(), WebCore::IntSize());
+    if (RunLoop::isMain())
+    {
+        page().handleMouseEvent(event);
+    }
+    else
+    {
+        RunLoop::main().dispatch([=]
+        {
+            handleMouseEvent(event);
+        });
+    }
+}
+
+void View::handleWheelEvent(const WebKit::NativeWebWheelEvent& event)
+{
+    if (RunLoop::isMain())
+    {
+        page().handleWheelEvent(event);
+    }
+    else
+    {
+        RunLoop::main().dispatch([=]
+        {
+            handleWheelEvent(event);
+        });
+    }
+}
+
+void View::handleTouchEvent(const WebKit::NativeWebTouchEvent& event)
+{
```

Comment:
There are no code duplication. Every function selects one of two alternate ways: do its own work (if called from a main thread) or recursively call itself from a main thread (otherwise).